feat: Add json schemas for manifest and manifest localization

daniel-schroeder-zeiss opened the pull request:
Adds schema files to this repository so that we can publish these via github. The contents are not yet final and may still be revised before the 1.0 release.

daniel-schroeder-zeiss committed:
feat: Add json schemas for manifest and manifest localization

czdietrich commented on schemas/manifest/schema.json:
```diff
@@ -0,0 +1,187 @@
+      "type": "string"
+    },
+    "documentation": {
+      "description": "The documentation url  for the plugin",
```
There 2 spaces between url and for.

czdietrich commented on schemas/manifest-localization/schema.json:
```diff
@@ -0,0 +1,46 @@
+          "description": "The localization for the plugin documentation",
+          "type": "string"
+        },
+        "<importAutomationId>.title": {
+          "description": "The localization for the plugin import automation title",
+          "type": "string"
+        },
+        "<importAutomationId>.description": {
+          "description": "The localization for the plugin import automation description",
+          "type": "string"
+        }
```
Are we ok with that syntax?
E.g. `<importAutomationId>.title`?

There seems to be something like "pattern properties", maybe we should use this instread?
https://json-schema.org/understanding-json-schema/reference/object#patternProperties

daniel-schroeder-zeiss replied:
Since we changed the provides object to directly present the provided module, we probably should just attribute the title and the summery to this provides object. This probably needs a code change in the AutoImporter though.

czdietrich replied:
Oh, here is something mixed up. I guess the latest implementation is to use:
`importAutomation.displayName`
`importAutomation.summary`
`importFormat.display`
So there is already a fixed ID for these keys.

daniel-schroeder-zeiss replied:
Yeah but I like the provides.summary and provides.displayName variant better.

czdietrich commented on schemas/manifest-localization/schema.json:
```diff
@@ -0,0 +1,46 @@
+          "description": "The localization for the plug-in documentation.",
+          "type": "string"
+        },
+        "provides.title": {
```
Should be `displayName` instead of `title` here.

daniel-schroeder-zeiss replied:
Fixed.

stephan-scholz-zeiss commented on schemas/manifest/schema.json:
```diff
@@ -0,0 +1,189 @@
+    "licenseName": {
+      "description": "The licence type of the plug-in",
+      "type": "string",
+      "enum": [
```
We should not provide a list of licenses, as this can probably never be complete.

daniel-schroeder-zeiss replied:
Yes, fixed.

daniel-schroeder-zeiss committed:
feat: Improve manifest schema and manifest localization schema

This contains the following improvements:
- consistent order of properties (description, type, format, ...)
- remove regular expressions where they are just an approximation or would change regularly
- add exact regular expression for sem ver
- descriptions are now sentences
- use anyof instead of oneof because it is much more performant and our
  rules exclude each other anyways
- add string type to file extension array and enforce unique entries
- add more restrictions to plugin ids
- do not restrict license names

daniel-schroeder-zeiss committed:
feat: Rename and move schema files to avoid unnecessary folder layer

stephan-scholz-zeiss commented on schemas/manifest-schema.json:
```diff
@@ -0,0 +1,147 @@
+      "type": "string"
+    },
+    "id": {
+      "title": "Plugin ID",
```
"Plug-in ID"

daniel-schroeder-zeiss replied:
Fixed.

stephan-scholz-zeiss commented on schemas/manifest-schema.json:
```diff
@@ -0,0 +1,147 @@
+          ],
+          "properties": {
+            "type": {
+              "description": "The plugin provides an import automation.",
```
Found another **plugin** ;)

daniel-schroeder-zeiss replied:
Fixed.

stephan-scholz-zeiss commented on schemas/manifest-schema.json:
```diff
@@ -0,0 +1,147 @@
+          ],
+          "properties": {
+            "type": {
+              "description": "The plugin provides an import format.",
```
And here. "plug-in"

daniel-schroeder-zeiss replied:
Fixed.

stephan-scholz-zeiss commented on schemas/manifest-schema.json:
```diff
@@ -0,0 +1,147 @@
+{
```
"manifestVersion" seems to be missing.

daniel-schroeder-zeiss replied:
Fixed.

daniel-schroeder-zeiss committed:
feat: Add manifest version to schema

czdietrich commented:
Should we set [additionalProperties: false](https://json-schema.org/understanding-json-schema/reference/object#additionalproperties) to disallow arbitrary properties in the manifest?

Currently it is possible to _accidentally_ type `"autor": "actually it's spelled author"` and won't see any squiggly lines.

czdietrich commented on schemas/manifest-schema.json:
```diff
@@ -0,0 +1,152 @@
+    "description",
+    "provides"
+  ],
+  "properties": {
```
Should we set [additionalProperties: false](https://json-schema.org/understanding-json-schema/reference/object#additionalproperties) to disallow arbitrary properties in the manifest?

Currently it is possible to _accidentally_ type `"autor": "actually it's spelled author"` and won't see any squiggly lines.

_This would be needed at several places though._

daniel-schroeder-zeiss replied:
Done.

czdietrich commented on schemas/manifest-schema.json:
```diff
@@ -0,0 +1,152 @@
+{
+  "$schema": "https://json-schema.org/draft/2020-12/schema",
```
Use schema specification that is supported by Visual Studio: `http://json-schema.org/draft-07/schema`.

daniel-schroeder-zeiss replied:
Done.

daniel-schroeder-zeiss committed:
feat: Lower schema version to be more compatible
